src/local_scheduler/local_scheduler_algorithm.cc

```diff
@@ -137,18 +148,15 @@ SchedulingAlgorithmState *SchedulingAlgorithmState_init(void) {
 
 void SchedulingAlgorithmState_free(SchedulingAlgorithmState *algorithm_state) {
   /* Free all of the tasks in the waiting queue. */
-  task_queue_entry *elt, *tmp1;
-  DL_FOREACH_SAFE(algorithm_state->waiting_task_queue, elt, tmp1) {
-    DL_DELETE(algorithm_state->waiting_task_queue, elt);
-    free(elt->spec);
-    free(elt);
+  for (auto &task : *algorithm_state->waiting_task_queue) {
+    TaskQueueEntry_free(&task);
```

Comment:
You should never have to explicitly free resources anywhere but inside a destructor in C++ code.
It seems like what you really want is `std::list<std::unique_ptr<TaskQueueEntry> >`, with the destructor `TaskQueueEntry::~TaskQueueEntry` defined appropriately. Then you can take out this `for` loop and just call `algorithm_state->waiting_task_queue.reset()` if needed.